src/JsonRpcEngine.ts

```diff
@@ -0,0 +1,350 @@
+      if (cb) {
+        this._handleBatch(req)
+          .then((res) => cb(null, res as JsonRpcResponse<unknown>[]))
+          .catch((err) => cb(err)); // fatal error
+        return undefined;
+      }
+      return this._handleBatch(req);
+    }
+
+    if (cb) {
+      return this._handle(req as JsonRpcRequest<unknown>, cb);
+    }
+    return this._promiseHandle(req as JsonRpcRequest<unknown>);
+  }
+
+  asMiddleware(): JsonRpcMiddleware {
+    return (req, res, next, end) => {
+      this._runAllMiddleware(req, res)
+        .then(async ({ isComplete, returnHandlers }) => {
+          if (isComplete) {
+            await this._runReturnHandlers(returnHandlers);
+            return end();
+          }
+
+          return next(async (handlerCallback) => {
+            try {
+              await this._runReturnHandlers(returnHandlers);
+            } catch (error) {
+              return handlerCallback(error);
+            }
+            return handlerCallback();
+          });
+        })
+        .catch((error) => {
+          end(error);
+        });
+    };
+  }
+
+  private async _handleBatch(
+    reqs: JsonRpcRequest<unknown>[],
+  ): Promise<JsonRpcResponse<unknown>[]> {
+    // The order here is important
+    // 3. Return batch response, or reject on some kind of fatal error
+    return await Promise.all(
+      // 2. Wait for all requests to finish
+      // 1. Begin executing each request in the order received
+      reqs.map(this._promiseHandle.bind(this)),
+    );
+  }
+
+  private _promiseHandle(
+    req: JsonRpcRequest<unknown>,
+  ): Promise<JsonRpcResponse<unknown>> {
+    return new Promise((resolve) => {
+      this._handle(req, (_err, res) => {
+        // There will always be a response, and it will always have any error
+        // that is caught and propagated.
+        resolve(res);
+      });
+    });
+  }
+
+  private _handle(
+    callerReq: JsonRpcRequest<unknown>,
+    cb: (
+      error: unknown,
+      response: JsonRpcResponse<unknown>
+    ) => void,
+  ): void {
+    const req: JsonRpcRequest<unknown> = { ...callerReq };
+    const res: InternalJsonRpcResponse = {
+      id: req.id,
+      jsonrpc: req.jsonrpc,
+    };
+
+    let processingError: Error | JsonRpcError;
+
+    this._processRequest(req, res)
+      .catch((error) => {
+        // either from return handlers or something unexpected
+        processingError = error;
+      })
+      .finally(() => {
+        // Preserve unserialized error, if any, for use in callback
+        const responseError = res._originalError;
+        delete res._originalError;
+
+        const error = responseError || processingError || null;
+
+        if (error) {
+          // Ensure no result is present on an errored response
+          delete res.result;
+          if (!res.error) {
+            res.error = serializeError(error);
+          }
+        }
+
+        cb(error, res as JsonRpcResponse<unknown>);
+      });
+  }
+
+  private async _processRequest(
```

Comment:
You could carry the generic all the way down the chain into these private methods, so long as you don't attempt to narrow this type they'll be treated as unknown but won't cause any heartache for end-users when they use the generics to define the params/response types.

Reply:
Might do that in the follow-up when I remove the internal types!